File: python/src/bleepstore/storage/sqlite.py
"""SQLite BLOB storage backend for BleepStore.

Implements the StorageBackend protocol using SQLite tables to store
object data and multipart part data as BLOBs in the same database
as metadata.

Tables:
    object_data(bucket, key, data, etag) — final assembled objects
    part_data(upload_id, part_number, data, etag) — multipart parts

This backend is useful for single-node deployments where keeping
everything in one SQLite database simplifies operations and backups.
"""

import hashlib
import logging
from collections.abc import AsyncIterator

import aiosqlite

logger = logging.getLogger(__name__)

# Streaming chunk size: 64 KB
_CHUNK_SIZE = 64 * 1024

_CREATE_OBJECT_DATA = """
CREATE TABLE IF NOT EXISTS object_data (
    bucket TEXT NOT NULL,
    key TEXT NOT NULL,
    data BLOB NOT NULL,
    etag TEXT NOT NULL,
    PRIMARY KEY (bucket, key)
)
"""

_CREATE_PART_DATA = """
CREATE TABLE IF NOT EXISTS part_data (
    upload_id TEXT NOT NULL,
    part_number INTEGER NOT NULL,
    data BLOB NOT NULL,
    etag TEXT NOT NULL,
    PRIMARY KEY (upload_id, part_number)
)
"""


class SQLiteStorageBackend:
    """Storage backend that persists object BLOBs inside a SQLite database.

    Objects are stored in the ``object_data`` table keyed by (bucket, key).
    Multipart parts are stored in the ``part_data`` table keyed by
    (upload_id, part_number).

    Attributes:
        db_path: Path to the SQLite database file.
    """

    def __init__(self, db_path: str) -> None:
        """Initialize the SQLite storage backend.

        Args:
            db_path: Path to the SQLite database file. This should typically
                     be the same database used for metadata storage.
        """
        self.db_path = db_path
        self._db: aiosqlite.Connection | None = None

    async def init(self) -> None:
        """Open the SQLite connection and create tables if they do not exist.

        Configures WAL mode and a 5-second busy timeout for concurrent access.
        """
        db = await aiosqlite.connect(self.db_path)
        self._db = db
        await db.execute("PRAGMA journal_mode=WAL")
        await db.execute("PRAGMA busy_timeout=5000")
        await db.execute(_CREATE_OBJECT_DATA)
        await db.execute(_CREATE_PART_DATA)
        await db.commit()
        logger.info("SQLite storage backend initialized at %s", self.db_path)

    async def close(self) -> None:
        """Close the SQLite connection."""
        if self._db is not None:
            await self._db.close()
            self._db = None

    def _ensure_db(self) -> aiosqlite.Connection:
        """Return the active database connection or raise."""
        if self._db is None:
            raise RuntimeError("SQLiteStorageBackend not initialized — call init() first")
        return self._db

    async def put(self, bucket: str, key: str, data: bytes) -> str:
        """Store an object's bytes in the database.

        Args:
            bucket: The bucket name.
            key: The object key.
            data: The raw bytes to store.

        Returns:
            The hex-encoded MD5 ETag of the stored data.
        """
        db = self._ensure_db()
        etag = hashlib.md5(data).hexdigest()
        await db.execute(
            "INSERT OR REPLACE INTO object_data (bucket, key, data, etag) VALUES (?, ?, ?, ?)",
            (bucket, key, data, etag),
        )
        await db.commit()
        return etag

    async def put_stream(
        self,
        bucket: str,
        key: str,
        stream: AsyncIterator[bytes],
        content_length: int | None = None,
    ) -> tuple[str, int]:
        """Stream-write an object by collecting chunks, then storing.

        Args:
            bucket: The bucket name.
            key: The object key.
            stream: Async iterator of byte chunks.
            content_length: Expected total size (optional, unused).

        Returns:
            A tuple of (hex-encoded MD5, total bytes written).
        """
        chunks: list[bytes] = []
        total = 0
        async for chunk in stream:
            chunks.append(chunk)
            total += len(chunk)

        data = b"".join(chunks)
        etag = await self.put(bucket, key, data)
        return etag, total

    async def put_part_stream(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_number: int,
        stream: AsyncIterator[bytes],
        content_length: int | None = None,
    ) -> tuple[str, int]:
        """Stream-write a multipart part by collecting chunks, then storing.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
            part_number: The sequential part number.
            stream: Async iterator of byte chunks.
            content_length: Expected total size (optional, unused).

        Returns:
            A tuple of (hex-encoded MD5, total bytes written).
        """
        chunks: list[bytes] = []
        total = 0
        async for chunk in stream:
            chunks.append(chunk)
            total += len(chunk)

        data = b"".join(chunks)
        etag = await self.put_part(bucket, key, upload_id, part_number, data)
        return etag, total

    async def get(self, bucket: str, key: str) -> bytes:
        """Retrieve an object's bytes from the database.

        Args:
            bucket: The bucket name.
            key: The object key.

        Returns:
            The raw bytes of the object.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        db = self._ensure_db()
        async with db.execute(
            "SELECT data FROM object_data WHERE bucket = ? AND key = ?",
            (bucket, key),
        ) as cursor:
            row = await cursor.fetchone()

        if row is None:
            raise FileNotFoundError(f"Object not found: {bucket}/{key}")

        return row[0]

    async def get_stream(
        self, bucket: str, key: str, offset: int = 0, length: int | None = None
    ) -> AsyncIterator[bytes]:
        """Retrieve an object's bytes as an async stream of 64 KB chunks.

        Args:
            bucket: The bucket name.
            key: The object key.
            offset: Byte offset to start reading from.
            length: Number of bytes to read, or None for all remaining.

        Yields:
            Chunks of bytes from the object.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        data = await self.get(bucket, key)

        # Apply offset
        if offset > 0:
            data = data[offset:]

        # Apply length limit
        if length is not None:
            data = data[:length]

        # Yield in chunks
        pos = 0
        while pos < len(data):
            end = min(pos + _CHUNK_SIZE, len(data))
            yield data[pos:end]
            pos = end

    async def delete(self, bucket: str, key: str) -> None:
        """Delete an object from the database.

        Silently succeeds if the object does not exist (idempotent).

        Args:
            bucket: The bucket name.
            key: The object key.
        """
        db = self._ensure_db()
        await db.execute(
            "DELETE FROM object_data WHERE bucket = ? AND key = ?",
            (bucket, key),
        )
        await db.commit()

    async def exists(self, bucket: str, key: str) -> bool:
        """Check if an object exists in the database.

        Args:
            bucket: The bucket name.
            key: The object key.

        Returns:
            True if the object exists.
        """
        db = self._ensure_db()
        async with db.execute(
            "SELECT 1 FROM object_data WHERE bucket = ? AND key = ?",
            (bucket, key),
        ) as cursor:
            row = await cursor.fetchone()

        return row is not None

    async def put_part(
        self, bucket: str, key: str, upload_id: str, part_number: int, data: bytes
    ) -> str:
        """Store a multipart upload part in the database.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
            part_number: The sequential part number.
            data: The raw bytes of this part.

        Returns:
            The hex-encoded MD5 ETag of the part.
        """
        db = self._ensure_db()
        etag = hashlib.md5(data).hexdigest()
        await db.execute(
            "INSERT OR REPLACE INTO part_data (upload_id, part_number, data, etag) VALUES (?, ?, ?, ?)",
            (upload_id, part_number, data, etag),
        )
        await db.commit()
        return etag

    async def assemble_parts(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_numbers: list[int],
    ) -> str:
        """Assemble uploaded parts into a final object.

        Reads each part from part_data in the specified order, concatenates
        them, computes the MD5, and inserts the result into object_data.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
            part_numbers: Ordered list of part numbers to assemble.

        Returns:
            The hex-encoded MD5 ETag of the assembled object.
        """
        db = self._ensure_db()
        assembled_chunks: list[bytes] = []

        for pn in part_numbers:
            async with db.execute(
                "SELECT data FROM part_data WHERE upload_id = ? AND part_number = ?",
                (upload_id, pn),
            ) as cursor:
                row = await cursor.fetchone()

            if row is None:
                raise FileNotFoundError(
                    f"Part not found: upload_id={upload_id}, part_number={pn}"
                )
            assembled_chunks.append(row[0])

        data = b"".join(assembled_chunks)
        etag = hashlib.md5(data).hexdigest()

        await db.execute(
            "INSERT OR REPLACE INTO object_data (bucket, key, data, etag) VALUES (?, ?, ?, ?)",
            (bucket, key, data, etag),
        )
        await db.commit()
        return etag

    async def delete_parts(self, bucket: str, key: str, upload_id: str) -> None:
        """Delete all stored parts for a multipart upload.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
        """
        db = self._ensure_db()
        await db.execute(
            "DELETE FROM part_data WHERE upload_id = ?",
            (upload_id,),
        )
        await db.commit()

    async def delete_upload_parts(self, upload_id: str) -> None:
        """Delete all stored parts for a multipart upload by upload ID only.

        Used during expired upload reaping where bucket/key are not needed
        since parts are keyed by upload_id in the database.

        Args:
            upload_id: The multipart upload identifier.
        """
        db = self._ensure_db()
        await db.execute(
            "DELETE FROM part_data WHERE upload_id = ?",
            (upload_id,),
        )
        await db.commit()

    async def copy_object(
        self,
        src_bucket: str,
        src_key: str,
        dst_bucket: str,
        dst_key: str,
    ) -> str:
        """Copy an object from one location to another within the database.

        Reads the source object and inserts (or replaces) at the destination.

        Args:
            src_bucket: The source bucket name.
            src_key: The source object key.
            dst_bucket: The destination bucket name.
            dst_key: The destination object key.

        Returns:
            The hex-encoded MD5 ETag of the copied object.

        Raises:
            FileNotFoundError: If the source object does not exist.
        """
        db = self._ensure_db()
        async with db.execute(
            "SELECT data FROM object_data WHERE bucket = ? AND key = ?",
            (src_bucket, src_key),
        ) as cursor:
            row = await cursor.fetchone()

        if row is None:
            raise FileNotFoundError(f"Source object not found: {src_bucket}/{src_key}")

        data = row[0]
        etag = hashlib.md5(data).hexdigest()

        await db.execute(
            "INSERT OR REPLACE INTO object_data (bucket, key, data, etag) VALUES (?, ?, ?, ?)",
            (dst_bucket, dst_key, data, etag),
        )
        await db.commit()
        return etag

File: golang/internal/storage/sqlite.go
package storage

import (
	"bytes"
	"context"
	"crypto/md5"
	"database/sql"
	"fmt"
	"io"
	"sort"

	_ "modernc.org/sqlite" // Pure-Go SQLite driver
)

// SQLiteBackend implements the StorageBackend interface using SQLite as the
// underlying data store. Object and part data are stored as BLOBs directly
// in the database, making this suitable for small-to-medium objects in
// single-node or embedded deployments.
type SQLiteBackend struct {
	db *sql.DB
}

// NewSQLiteBackend creates a new SQLiteBackend backed by the given database
// file path. It opens the database, applies performance PRAGMAs, and creates
// the required tables.
func NewSQLiteBackend(dbPath string) (*SQLiteBackend, error) {
	db, err := sql.Open("sqlite", dbPath)
	if err != nil {
		return nil, fmt.Errorf("opening SQLite storage database: %w", err)
	}

	b := &SQLiteBackend{db: db}
	if err := b.initDB(); err != nil {
		db.Close()
		return nil, fmt.Errorf("initializing SQLite storage database: %w", err)
	}
	return b, nil
}

// initDB applies PRAGMAs and creates the required tables.
func (b *SQLiteBackend) initDB() error {
	pragmas := []string{
		"PRAGMA journal_mode = WAL",
		"PRAGMA busy_timeout = 5000",
	}
	for _, p := range pragmas {
		if _, err := b.db.Exec(p); err != nil {
			return fmt.Errorf("executing %q: %w", p, err)
		}
	}

	schema := `
		CREATE TABLE IF NOT EXISTS object_data (
			bucket TEXT NOT NULL,
			key    TEXT NOT NULL,
			data   BLOB NOT NULL,
			etag   TEXT NOT NULL,
			PRIMARY KEY (bucket, key)
		);

		CREATE TABLE IF NOT EXISTS part_data (
			upload_id   TEXT    NOT NULL,
			part_number INTEGER NOT NULL,
			data        BLOB    NOT NULL,
			etag        TEXT    NOT NULL,
			PRIMARY KEY (upload_id, part_number)
		);
	`
	if _, err := b.db.Exec(schema); err != nil {
		return fmt.Errorf("creating storage schema: %w", err)
	}
	return nil
}

// Close closes the underlying SQLite database connection.
func (b *SQLiteBackend) Close() error {
	if b.db != nil {
		return b.db.Close()
	}
	return nil
}

// PutObject reads all data from the reader, computes an MD5 ETag, and stores
// the data as a BLOB in the object_data table. Uses INSERT OR REPLACE so that
// re-uploads overwrite the existing row.
func (b *SQLiteBackend) PutObject(ctx context.Context, bucket, key string, reader io.Reader, size int64) (int64, string, error) {
	data, err := io.ReadAll(reader)
	if err != nil {
		return 0, "", fmt.Errorf("reading object data: %w", err)
	}

	etag := computeETag(data)

	_, err = b.db.ExecContext(ctx,
		`INSERT OR REPLACE INTO object_data (bucket, key, data, etag) VALUES (?, ?, ?, ?)`,
		bucket, key, data, etag,
	)
	if err != nil {
		return 0, "", fmt.Errorf("putting object %q/%q: %w", bucket, key, err)
	}

	return int64(len(data)), etag, nil
}

// GetObject retrieves the object data and ETag from the object_data table.
// Returns an io.NopCloser wrapping a bytes.Reader for the data, the data
// size, the ETag, and any error. Returns an error if the object is not found.
func (b *SQLiteBackend) GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, string, error) {
	var data []byte
	var etag string

	err := b.db.QueryRowContext(ctx,
		`SELECT data, etag FROM object_data WHERE bucket = ? AND key = ?`,
		bucket, key,
	).Scan(&data, &etag)
	if err == sql.ErrNoRows {
		return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
	}
	if err != nil {
		return nil, 0, "", fmt.Errorf("getting object %q/%q: %w", bucket, key, err)
	}

	return io.NopCloser(bytes.NewReader(data)), int64(len(data)), etag, nil
}

// DeleteObject removes the object data from the object_data table.
// Idempotent: deleting a non-existent object is not an error.
func (b *SQLiteBackend) DeleteObject(ctx context.Context, bucket, key string) error {
	_, err := b.db.ExecContext(ctx,
		`DELETE FROM object_data WHERE bucket = ? AND key = ?`,
		bucket, key,
	)
	if err != nil {
		return fmt.Errorf("deleting object %q/%q: %w", bucket, key, err)
	}
	return nil
}

// CopyObject copies an object from the source bucket/key to the destination
// bucket/key by reading the source data and inserting it at the destination.
// Returns the ETag of the copied object (which is the same as the source).
func (b *SQLiteBackend) CopyObject(ctx context.Context, srcBucket, srcKey, dstBucket, dstKey string) (string, error) {
	var data []byte
	var etag string

	err := b.db.QueryRowContext(ctx,
		`SELECT data, etag FROM object_data WHERE bucket = ? AND key = ?`,
		srcBucket, srcKey,
	).Scan(&data, &etag)
	if err == sql.ErrNoRows {
		return "", fmt.Errorf("source object not found: %s/%s", srcBucket, srcKey)
	}
	if err != nil {
		return "", fmt.Errorf("reading source object %q/%q: %w", srcBucket, srcKey, err)
	}

	_, err = b.db.ExecContext(ctx,
		`INSERT OR REPLACE INTO object_data (bucket, key, data, etag) VALUES (?, ?, ?, ?)`,
		dstBucket, dstKey, data, etag,
	)
	if err != nil {
		return "", fmt.Errorf("writing destination object %q/%q: %w", dstBucket, dstKey, err)
	}

	return etag, nil
}

// PutPart writes a single part of a multipart upload into the part_data table.
// Uses INSERT OR REPLACE so re-uploading the same part number overwrites the
// previous data.
func (b *SQLiteBackend) PutPart(ctx context.Context, bucket, key, uploadID string, partNumber int, reader io.Reader, size int64) (string, error) {
	data, err := io.ReadAll(reader)
	if err != nil {
		return "", fmt.Errorf("reading part data: %w", err)
	}

	etag := computeETag(data)

	_, err = b.db.ExecContext(ctx,
		`INSERT OR REPLACE INTO part_data (upload_id, part_number, data, etag) VALUES (?, ?, ?, ?)`,
		uploadID, partNumber, data, etag,
	)
	if err != nil {
		return "", fmt.Errorf("putting part %d for upload %q: %w", partNumber, uploadID, err)
	}

	return etag, nil
}

// AssembleParts concatenates the specified parts into a single object and stores
// the result in the object_data table. The composite ETag is computed by hashing
// the concatenation of the individual part MD5 digests, producing the standard
// S3 multipart ETag format: "md5-of-part-md5s-N".
func (b *SQLiteBackend) AssembleParts(ctx context.Context, bucket, key, uploadID string, partNumbers []int) (string, error) {
	// Sort part numbers to ensure correct assembly order.
	sorted := make([]int, len(partNumbers))
	copy(sorted, partNumbers)
	sort.Ints(sorted)

	var assembled bytes.Buffer
	compositeMD5 := md5.New()

	for _, pn := range sorted {
		var data []byte
		err := b.db.QueryRowContext(ctx,
			`SELECT data FROM part_data WHERE upload_id = ? AND part_number = ?`,
			uploadID, pn,
		).Scan(&data)
		if err == sql.ErrNoRows {
			return "", fmt.Errorf("part %d not found for upload %q", pn, uploadID)
		}
		if err != nil {
			return "", fmt.Errorf("reading part %d for upload %q: %w", pn, uploadID, err)
		}

		assembled.Write(data)

		// Compute MD5 of this individual part and feed it into the composite hash.
		partHash := md5.Sum(data)
		compositeMD5.Write(partHash[:])
	}

	// Composite ETag format: "md5-of-concatenated-part-md5s-N"
	etag := fmt.Sprintf(`"%x-%d"`, compositeMD5.Sum(nil), len(sorted))

	// Store the assembled object.
	_, err := b.db.ExecContext(ctx,
		`INSERT OR REPLACE INTO object_data (bucket, key, data, etag) VALUES (?, ?, ?, ?)`,
		bucket, key, assembled.Bytes(), etag,
	)
	if err != nil {
		return "", fmt.Errorf("storing assembled object %q/%q: %w", bucket, key, err)
	}

	// Clean up parts for this upload.
	_, err = b.db.ExecContext(ctx,
		`DELETE FROM part_data WHERE upload_id = ?`,
		uploadID,
	)
	if err != nil {
		return "", fmt.Errorf("cleaning up parts for upload %q: %w", uploadID, err)
	}

	return etag, nil
}

// DeleteParts removes all part data associated with the given multipart upload.
func (b *SQLiteBackend) DeleteParts(ctx context.Context, bucket, key, uploadID string) error {
	_, err := b.db.ExecContext(ctx,
		`DELETE FROM part_data WHERE upload_id = ?`,
		uploadID,
	)
	if err != nil {
		return fmt.Errorf("deleting parts for upload %q: %w", uploadID, err)
	}
	return nil
}

// DeleteUploadParts removes all part data for a specific upload ID. This is
// used during startup reaping of expired uploads to clean up orphaned part
// data in the database.
func (b *SQLiteBackend) DeleteUploadParts(uploadID string) error {
	_, err := b.db.Exec(
		`DELETE FROM part_data WHERE upload_id = ?`,
		uploadID,
	)
	if err != nil {
		return fmt.Errorf("deleting upload parts for %q: %w", uploadID, err)
	}
	return nil
}

// CreateBucket is a no-op for the SQLite backend because bucket organization
// is handled entirely through the (bucket, key) composite keys in the tables.
// The metadata layer tracks bucket existence.
func (b *SQLiteBackend) CreateBucket(ctx context.Context, bucket string) error {
	return nil
}

// DeleteBucket is a no-op for the SQLite backend. Object data rows are deleted
// individually via DeleteObject. The metadata layer manages bucket lifecycle.
func (b *SQLiteBackend) DeleteBucket(ctx context.Context, bucket string) error {
	return nil
}

// ObjectExists checks whether an object exists in the object_data table.
func (b *SQLiteBackend) ObjectExists(ctx context.Context, bucket, key string) (bool, error) {
	var count int
	err := b.db.QueryRowContext(ctx,
		`SELECT COUNT(*) FROM object_data WHERE bucket = ? AND key = ?`,
		bucket, key,
	).Scan(&count)
	if err != nil {
		return false, fmt.Errorf("checking object existence %q/%q: %w", bucket, key, err)
	}
	return count > 0, nil
}

// HealthCheck verifies that the SQLite storage database is operational by
// executing a simple query.
func (b *SQLiteBackend) HealthCheck(ctx context.Context) error {
	var n int
	return b.db.QueryRowContext(ctx, `SELECT 1`).Scan(&n)
}
